statsmodels/base/data.py

```diff
@@ -333,10 +333,12 @@ def drop_nans_2d(x):
     def _convert_endog_exog(self, endog, exog):
 
         # for consistent outputs if endog is (n,1)
-        yarr = self._get_yarr(endog)
+        # We call __array__() to convert to an array if the object is array-like
+        # but not yet an array. For actual ndarrays, this does nothing.
+        yarr = self._get_yarr(endog.__array__())
```

Comment:
How far backward compatabile is this?  Should we try excep to improve backward compay?

Reply:
Based on only partial code skimming

AFAICS
endog here could still be a list or other non-array like data structure.
The current code handles anything that can be `np.asarray`ed.
There is not really any special handling of arrays as user input, Internally all computations are after `asarray`,  

The main point of handling Dataframes separately is to keep track of the row index and column names.

Reply:
> endog here could still be a list or other non-array data structure.

Ok, that could indeed be because I did not include lists in the tests so far. That should be properly tested.